tstest/mts/mts.go

```diff
@@ -0,0 +1,287 @@
+		out, err := exec.Command("go", "list", "-f", "{{.Target}}", "tailscale.com/cmd/tailscaled").CombinedOutput()
+		if err != nil {
+			return "", err
+		}
+		return strings.TrimSpace(string(out)), nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	return v
+}
+
+func (s *Server) Run() {
+	if err := os.MkdirAll(instDir(), 0700); err != nil {
+		panic(err)
+	}
+
+	log.Printf("Running all Tailscale daemons: %q", s.InstanceNames())
+	for _, name := range s.InstanceNames() {
+		go s.RunInstance(name)
+	}
+
+	sock := mtsSock()
+	os.Remove(sock)
+	log.Printf("Listening on %q ...", sock)
+	ln, err := net.Listen("unix", sock)
+	if err != nil {
+		panic(err)
+	}
+	log.Fatal(http.Serve(ln, s))
+}
+
+var validNameRx = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)
+
+func validInstanceName(name string) bool {
+	return validNameRx.MatchString(name)
+}
+
+func (s *Server) InstanceRunning(name string) bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, ok := s.cmds[name]
+	return ok
+}
+
+func (s *Server) RunInstance(name string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.cmds[name]; ok {
+		return fmt.Errorf("instance %q already running", name)
+	}
+
+	if !validInstanceName(name) {
+		panic(fmt.Sprintf("invalid instance name %q", name))
+	}
+	dir := filepath.Join(instDir(), name)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return err
+	}
+	log.Printf("Running Tailscale daemon %q in %q", name, dir)
+
+	cmd := exec.Command(s.tailscaled(),
+		"--tun=userspace-networking",
+		"--statedir="+filepath.Join(dir),
+		"--socket="+filepath.Join(dir, "tailscaled.sock"),
+		"--verbose=1",
+	)
+	cmd.Dir = dir
+
+	// TODO(bradfitz): capture these, record in memory, serve
+	// them via HTTP?
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	go func() {
+		err := cmd.Wait()
+		log.Printf("Tailscale daemon %q exited: %v", name, err)
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		delete(s.cmds, name)
+	}()
+
+	mak.Set(&s.cmds, name, cmd)
+	return nil
+}
+
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if inst, ok := strings.CutPrefix(r.URL.Path, "/create/"); ok {
+		if !s.InstanceRunning(inst) {
+			if err := s.RunInstance(inst); err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+		}
+		fmt.Fprintf(w, "OK\n")
+		return
+	}
+	fmt.Fprintf(w, "Hello, %s\n", r.URL.Path)
+}
+
+func (s *Server) InstanceNames() []string {
+	var ret []string
+	des, err := os.ReadDir(instDir())
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		panic(err)
+	}
+	for _, de := range des {
+		if !de.IsDir() {
+			continue
+		}
+		ret = append(ret, de.Name())
+	}
+	return ret
+}
+
+func instDir() string {
+	dir, err := os.UserConfigDir()
+	if err != nil {
+		panic(err)
+	}
+	return filepath.Join(dir, "multi-tailscale-dev")
+}
+
+func instSock(name string) string {
+	return filepath.Join(instDir(), name, "tailscaled.sock")
```

Comment:
I'm slightly sad once again that Go doesn't have os.RuntimeDir, which is where these should really live, rather than in ~/.config.